src/callbacks_step/save_solution_covariant.jl

```diff
@@ -0,0 +1,106 @@
+# Convert to another set of variables using a solution_variables function that depends on 
+# auxiliary variables
+function convert_variables(u, solution_variables, mesh::P4estMesh{2},
+                           equations, dg, cache)
+    (; aux_node_vars) = cache.auxiliary_variables
+
+    # Extract the number of solution variables to be output 
+    # (may be different than the number of conservative variables) 
+    n_vars = length(solution_variables(Trixi.get_node_vars(u, equations, dg, 1, 1, 1),
+                                       get_node_aux_vars(aux_node_vars, equations, dg, 1, 1,
+                                                         1), equations))
+    # Allocate storage for output
+    data = Array{eltype(u)}(undef, n_vars, nnodes(dg), nnodes(dg), nelements(dg, cache))
+
+    # Loop over all nodes and convert variables, passing in auxiliary variables
+    Trixi.@threaded for element in eachelement(dg, cache)
+        for j in eachnode(dg), i in eachnode(dg)
+            u_node = Trixi.get_node_vars(u, equations, dg, i, j, element)
+            a_node = get_node_aux_vars(aux_node_vars, equations, dg, i, j, element)
+            u_node_converted = solution_variables(u_node, a_node, equations)
+            for v in 1:n_vars
+                data[v, i, j, element] = u_node_converted[v]
+            end
+        end
+    end
+    return data
+end
+
+# Specialized save_solution_file method that supports a solution_variables function 
```

Comment:
Instead of re-implementing the entire function, couldn't you just
1) perform the conversion `u` --> `data` using the aux vars
2) call the original `Trixi.save_solution_file` with `data` as the argument for `u` and `cons2cons` as the argument for `solution_variables`
? If this works, it would save you a lot of code duplication...

Reply:
That is a good suggestion, but it did not quite work due to the fact that `cons2cons` requires the number of variables to be output to equal `nvariables(equations)` (the `varnames` would also not match). However, I was able to implement a bit of a hack to fix this by requiring two `solution_variables` methods to be defined:

`solution_variables(u, aux_vars equations) = ...` (does the actual conversion)
`solution_variables(u, equations) = u` (does nothing)  

The first is called within my new `convert_variables` function, and then the resulting`data` is passed into the existing `Trixi.save_solution_file`, which uses the second method (i.e. the no-op) on the already converted data.

This removes the repeated code as well as the dependency on HDF5.

Reply:
I forgot that this line also requires `solution_variables` to return an `SVector` of length `nvariables(equations)`:

https://github.com/trixi-framework/Trixi.jl/blob/824f7a8a5e433c8deb29245bf9a79c3558bce397/src/callbacks_step/save_solution_dg.jl#L36-L39

This is not a big deal for me right now, so perhaps it's not worth adding back in the duplicated code to this PR, but the ability to output an arbitrary number of variables is a feature I'd eventually like to have.